File: 2017/day21/day21.py
#!/usr/bin/env python3
# https://adventofcode.com/2017/day/21

import sys
import typing as t
from copy import deepcopy
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
self_tests = "-T" in sys.argv
if self_tests:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()


Square = t.List[t.List]


def subsquare(grid: Square, x: int, y: int, n: int) -> Square:
    return list(list(grid[y + j][x + i] for i in range(n)) for j in range(n))


def rotate(square: Square) -> None:
    n = len(square)
    orig = deepcopy(square)
    for x in range(n):
        for y in range(n):
            square[y][x] = orig[n - 1 - x][y]


def flip(square: Square) -> None:
    n = len(square)
    for y in range(n):
        for x in range(n // 2):
            square[y][x], square[y][n - 1 - x] = square[y][n - 1 - x], square[y][x]


def iter_pos(grid: Square, x: int, y: int, n: int):
    square = subsquare(grid, x, y, n)
    for _ in range(4):
        rotate(square)
        yield square

    flip(square)
    yield square

    for _ in range(3):
        rotate(square)
        yield square


def pattern(square: Square) -> str:
    return "/".join("".join(row) for row in square)


def show(grid) -> None:
    print("\n".join("".join(row) for row in grid))
    print()


def enhance(rules: t.Dict[str, Square], grid: Square) -> Square:
    n = len(grid)

    if n % 2 == 0:
        m = 2
    elif n % 3 == 0:
        m = 3
    else:
        assert False

    nn = (n // m) * (m + 1)
    ng = [[None for _ in range(nn)] for _ in range(nn)]

    for y in range(n // m):
        for x in range(n // m):
            square = subsquare(grid, m * x, m * y, m)
            enhancement = rules[pattern(square)]
            for j in range(m + 1):
                for i in range(m + 1):
                    ng[y * (m + 1) + j][x * (m + 1) + i] = enhancement[j][i]

    return ng


def pixels(grid: Square) -> int:
    return sum(sum(1 for c in row if c == "#") for row in grid)


# parse the rules and build a dictionaru for all rotates/flipped rules

rules = {}

for line in lines:
    a, b = line.split(" => ", maxsplit=1)

    a = list(a.split("/"))
    b = list(b.split("/"))

    for e in iter_pos(a, 0, 0, len(a)):
        rules[pattern(e)] = b


# set the initial grid

grid = [list(".#."), list("..#"), list("###")]


# enhance the grid according to the rules

if filename == "test.txt":
    for _ in range(2):
        grid = enhance(rules, grid)
    print(pixels(grid))
    show(grid)

    assert pixels(grid) == 12

else:
    # part 1
    for _ in range(5):
        grid = enhance(rules, grid)
    print(pixels(grid))

    # part 2
    for _ in range(13):
        grid = enhance(rules, grid)
    print(pixels(grid))

File: 2017/day21/day21.rs
//! [Day 21: Fractal Art](https://adventofcode.com/2017/day/21)

use rustc_hash::FxHashMap;

type Square = aoc::Square<u8>;

struct Puzzle {
    rules: FxHashMap<Square, Square>,
    start: Square,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        let mut p = Self {
            rules: FxHashMap::default(),
            start: Square::parse(".#./..#/###", '/'),
        };

        for line in data.lines() {
            let (src, dest) = line.split_once(" => ").unwrap();

            let src = Square::parse(src, '/');
            let dest = Square::parse(dest, '/');

            for s in src.iter_pos() {
                p.rules.insert(s, dest.clone());
            }
        }
        p
    }

    fn enhance(&self, grid: &Square) -> Square {
        let n = grid.size();

        let m = if n % 2 == 0 {
            2
        } else if n % 3 == 0 {
            3
        } else {
            panic!();
        };

        let new_n = (n / m) * (m + 1);
        let mut enhanced_grid = Square::new(new_n);

        for y in 0..(n / m) {
            for x in 0..(n / m) {
                let square = grid.get_square(m * x, m * y, m);
                let enhancement = &self.rules[&square];

                enhanced_grid.put_square(x * (m + 1), y * (m + 1), enhancement);
            }
        }

        enhanced_grid
    }

    fn solve(&self, iterations: usize) -> usize {
        let mut grid = self.start.clone();

        for _ in 0..iterations {
            grid = self.enhance(&grid);
        }

        bytecount::count(grid.values(), b'#')
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        self.solve(5)
    }

    /// Solve part two.
    fn part2(&self) -> usize {
        self.solve(18)
    }
}

fn main() {
    let args = aoc::parse_args();
    let puzzle = Puzzle::new(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part1() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.solve(2), 12);
    }
}
